app/models/concerns/course/settings/email_settings_concern.rb

```diff
@@ -0,0 +1,103 @@
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    # Email setting items for the given component.
+    #
+    # @return [Hash] Each key represents an email setting. Each value is a Hash
+    #   representing defaults for the email setting, e.g.
+    #     {
+    #       item_opening: { enabled_by_default: false },
+    #       item_closing: { enabled_by_default: true, days_in_advance: 2 },
+    #     }
+    def email_setting_items
+      raise NotImplementedError, 'Define component email settings in its settings model.'
+    end
+
+    # Returns the class of the component which the setting model is associated with.
+    #
+    # @return [Class] The component class
+    def component_class
+      raise NotImplementedError, 'Specify component class in settings model.'
+    end
+
+    # Checks whether a type of email notification is enabled for a course.
+    #
+    # @param [Course] course
+    # @param [Symbol] key The email notification type
+    def email_enabled?(course, key)
+      raise ArgumentError, 'Invalid email key' unless valid_email_setting_key?(key)
+
+      user_setting = course.settings(component_class.key, :emails, key).enabled
+      user_setting.nil? ? email_setting_items[key][:enabled_by_default] : user_setting
+    end
+
+    # Updates a single email setting.
+    #
+    # @param [#settings] settings
+    # @param [Symbol] key Key of the email notification setting to update
+    # @param [Boolean] value Value of the setting to update to
+    # @return [Boolean] true if updating succeeds, false otherwise
+    def update_email_setting(settings, key, value)
+      return false unless valid_email_setting_key?(key)
+      settings.settings(:emails, key).enabled = value
+      true
+    end
+
+    private
+
+    def valid_email_setting_key?(key)
+      email_setting_items.key?(key)
+    end
+  end
+
+  delegate :email_setting_items, to: :class
+  delegate :component_class, to: :class
+
+  # A list of concrete email settings for the component. This is used by
+  # {Course::Settings::Notifications} for the notifications settings page.
+  # See {Course::Settings::Notifications#email_settings} for details of the hash shape.
+  #
+  # @return [Array<Hash>]
+  def email_settings
+    email_setting_items.map do |key, defaults|
+      user_setting = @settings.settings(:emails, key).enabled
+      enabled = user_setting.nil? ? defaults[:enabled_by_default] : user_setting
+      setting = { component: component_class.key, key: key, enabled: enabled }
+      respond_to?(:title) && title ? setting.merge(component_title: title) : setting
+    end
+  end
+
+  # Updates a single email setting.
+  #
+  # @param [Hash] attributes Hash with `'key'` and `'enabled'` keys, e.g.
+  #   { 'key' => 'item_opening', 'enabled' => true }
+  # @return [Boolean] true if updating succeeds, false otherwise
+  def update_email_setting(attributes)
+    self.class.update_email_setting(@settings, attributes['key'].to_sym, attributes['enabled'])
```

Comment:
Okay, I think I am confused because both class and instance methods are public, not sure which one to use.  We should make non-API methods private. 

Actually you can use `OpenStruct` to define a dummy context: `context = OpenStuct.new(current_course: xxx, key: xxx)` to make the test easier.